### PR TITLE
Add scene preview panel

### PR DESCRIPTION
**Pull request by DaelonSuzuka**
This adds a panel to the activity bar that shows a tree preview of the currently selected scene file.

https://user-images.githubusercontent.com/18042232/186282657-50ba98f8-f8c1-49ae-80f3-4846aafd4fe2.mp4

Things that need fixing in this PR:

- [x] find a Godot icon svg for the activity bar button
- [x] there's a minor bug where it throws an error popup when switching to some non-code editors (specifically the "running extensions" page)
- [x] don't clear the preview when selecting a non-scene file. this will allow looking at the structure of a scene while writing code in a gdscript file

~~future improvements (not this PR):~~
j/k these are all in this PR now:

- [x] add context menu to the treeitems to allow things like copying the relative NodePath of given item
- [x] when clicking on treeitems, scroll to/display the associated section of the scene file
- [x] button to allow manually pinning the displayed scene
- [x] parse other data from the scene file and display more things in the preview:
  - attached scripts
  - [x] nodes marked as Unique (for name copying purposes)
  - collapsible section for external resources and/or subresources

### CODE REVIEW

**Comment by atirut-w** [comment]
I think this should be in the document outline but alas, guess the doc outline is an LSP thing.

**Comment by DaelonSuzuka** [comment]
> I think this should be in the document outline but alas, guess the doc outline is an LSP thing.

I don't understand how the two are related at all.

**Comment by atirut-w** [comment]
> I don't understand how the two are related at all.

I am talking about the outline section under the document tab. Sounds like a good place to put the scene tree preview in.

**Comment by DaelonSuzuka** [comment]
Oh, in the Explorer section in the Primary Sidebar? I interpreted it as the documentation pages that are provided by the LSP.

Yeah, this is basically a document outline for a scene file, so I can see the argument for putting it there. Personally, that's a useless location since if I want to see files, I *only* want to see files. I always hide Open Editors, Timeline, Outline, anything that's in that category except the file Explorer.

Luckily, you can just drag any sidebar containers to any section you want, so it would be one mouseclick to move it there if that's what you like.

In addition to "because I don't want it there" (which I will admit is not a great reason!), I also think the default location should be its own category to improve user discoverability. If the default location for this thing is minimized at the bottom of the Explorer tab, how many people will even notice that it was added? Adding a Godot category provides an immediately visible, obvious default location for this and other features that I hope I can work on soon.

By the way, screenshots of the new Godot icon:

ShareX_UmRWhXzotA

**Comment by DaelonSuzuka** [comment]
I figured the easiest way to explain the new features was with a series of short video clips:

<details>
<summary>Scene Preview Pinning</summary>

Clicking the pushpin button locks the currently displayed scene. This makes it easier to use the Scene Preview as a reference while working on scripts.

https://user-images.githubusercontent.com/18042232/187405484-23d2d276-df3c-45be-89a1-5df363892838.mp4

</details>

<details>
<summary>Switch between a scene and a related script</summary>

In C/C++, pressing `alt+o` will switch from `foo.c` to `foo.h` and vice-versa. I replicated the same functionality between a script and a related scene. 

https://user-images.githubusercontent.com/18042232/187405998-ec54168e-d959-4e58-a43d-1927e2ac9040.mp4

</details>

<details>
<summary>Display Scene Preview when viewing related script</summary>

Using the same mechanism as above, the Scene Preview will show you the scene file that matches the script you have open.

https://user-images.githubusercontent.com/18042232/187409272-c1104d6d-875f-45cc-9ad5-d6930dcbb024.mp4

</details>

<details>
<summary>Tooltips with syntax highlighting</summary>

Hovering over a node in the scene tree will now show you a tooltip of that node's entire definition.

https://user-images.githubusercontent.com/18042232/187406391-cdaeba42-3e46-4f5b-aee9-951d210f6012.mp4

</details>


Nodes in the Scene Preview now have several context menu options:

<details>
<summary>Right Click > Copy Node Path</summary>

This menu option will copy the path of the selected node, relative to the root of the current scene. If the node has been marked as Unique, the unique node path will be copied instead (`%NodeName`).

https://user-images.githubusercontent.com/18042232/187405535-482146a7-a791-4582-80b6-f6f6fe730d62.mp4

</details>

<details>
<summary>Right Click > Go to Definition</summary>

This menu option will jump to and highlight the definition of the selected node.

https://user-images.githubusercontent.com/18042232/187407085-ee616f24-aac4-48d6-8d24-edd5f5c528fb.mp4

</details>

<details>
<summary>Right Click > Open Scene</summary>

This menu option is only available on PackedScene nodes, aka instances of other scenes. Selecting this will open the scene file that the selected node is an instance of.

https://user-images.githubusercontent.com/18042232/187407398-d9cdd907-36e1-4dde-9192-6be3ec1dd7a5.mp4

</details>

**Comment by DaelonSuzuka** [comment]
I will probably add a couple more settings, to let the user tweak some of these advanced behaviors.

In the same vein, it's probably appropriate to add some commands to allow manually invoking features.

I very desperately wanted to put icons on the far right of each tree item to indicate it's visibility, whether it has a script attached, if it's been marked unique, etc etc. For some blasted reason, this appears to be literally impossible with the current vscode TreeView API. It's possible to add *buttons* that are hidden until you move your mouse over them, but there's no way to get permanently visible icons like in the Godot Editor's Scene Tree.

Something like this is probably the best I'll be able to do:
image

**Comment by DaelonSuzuka** [comment]
I added a configuration that will let the user pick how aggressively the Scene Preview will search for a "related scene" when opening a script. The options are:
- `anyFolder`: Attempt to preview a related scene from anywhere in the workspace.
- `sameFolder`: Attempt to preview a related scene from the same folder.
- `off`: Do not attempt to preview a related scene.

The default setting is `sameFolder`, because I figure that will be helpful and relatively intuitive(aka, not magic), and has a limited chance of grabbing a totally unrelated scene.

I also figured out a relatively clean way to show more information about a given Node. Right now it's just whether there's a script attached and if the node is Unique, but I can add more later if something seems useful.

image

**Comment by DaelonSuzuka** [comment]
I think I'm done with this for now.

File: package.json [hunk 1]
```diff
@@ -184,6 +225,20 @@
 					],
 					"default": "beside",
 					"description": "Where to place the native symbol windows"
+				},
+                "godot_tools.scenePreview.previewRelatedScenes": {
```
Comment (Calinou) [inline package.json:229]: Mixed indentation – please use tabs in the whole file.

File: package.json [hunk 2]
```diff
@@ -184,6 +225,20 @@
...
+					"enum": [
+						"anyFolder",
+						"sameFolder",
+						"off"
+					],
+					"enumDescriptions": [
+						"Attempt to preview a related scene from anywhere in the workspace.",
+						"Attempt to preview a related scene from the same folder.",
+                        "Do not attempt to preview a related scene."
```
Comment (Calinou) [inline package.json:238]: Same as above.

File: src/godot-tools.ts [hunk 3]
```diff
@@ -120,6 +125,21 @@ export class GodotTools {
 		this.client.open_documentation(symbolName);
 	}
 
+    private async switch_scene_script() {
```
Comment (Calinou) [inline src/godot-tools.ts:128]: Use tabs for indentation.

File: src/scene_preview_provider.ts [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+    private externalResources = {};
+
+	private changeEvent = new EventEmitter<void>();
+
+	public get onDidChangeTreeData(): Event<void> {
+		return this.changeEvent.event;
+	}
+
+	public async refresh() {
+        if (this.scenePreviewPinned) {
```
Comment (Calinou) [inline src/scene_preview_provider.ts:34]: Use tabs for indentation in that file too.

File: src/utils.ts [hunk 5]
```diff
@@ -13,3 +15,29 @@ export function set_configuration(name: string, value: any) {
 export function is_debug_mode(): boolean {
 	return process.env.VSCODE_DEBUG_MODE === "true";
 }
+
+export function set_context(name: string, value: any) {
+    vscode.commands.executeCommand('setContext', name, value);
```
Comment (Calinou) [inline src/utils.ts:20]: Use tabs for indentation.

File: src/utils.ts [hunk 6]
```diff
@@ -1,4 +1,6 @@
 import * as vscode from "vscode";
+import * as path from 'path';
```
Comment (Calinou) [inline src/utils.ts:2]: Use double quotes for strings in that file.

File: src/godot-tools.ts [hunk 7]
```diff
@@ -50,8 +52,11 @@ export class GodotTools {
 		vscode.commands.registerCommand("godot-tool.copy_resource_path_context", this.copy_resource_path.bind(this));
 		vscode.commands.registerCommand("godot-tool.copy_resource_path", this.copy_resource_path.bind(this));
 		vscode.commands.registerCommand("godot-tool.open_type_documentation", this.open_type_documentation.bind(this));
+		vscode.commands.registerCommand("godotTools.switchSceneScript", this.switch_scene_script.bind(this));
 
-		vscode.commands.executeCommand('setContext', 'godotTools.connectedToEditor', false);
+		set_context('godotTools.context.connectedToEditor', false);
```
Comment (Calinou) [inline src/godot-tools.ts:57]: Use double quotes for strings in that file.

**Review by Calinou — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (DaelonSuzuka) [reply package.json:229]: I want it on the record that I did spend several hours trying to force Prettier to format to our style, and all I accomplished was getting it to stop working entirely.

**Comment by DaelonSuzuka** [comment]
I want it on the record that I did spend several hours trying to convince Prettier to format to our style, and all I accomplished was getting it to stop working entirely.

**Comment by Calinou** [comment]
> I want it on the record that I did spend several hours trying to convince Prettier to format to our style, and all I accomplished was getting it to stop working entirely.

I'm OK with changing the formatting to be more standard across what other popular VS Code extensions are doing. This should be done in a future PR though (and after all prominent PRs are merged).

**Comment by DaelonSuzuka** [comment]
> I'm OK with changing the formatting to be more standard across what other popular VS Code extensions are doing. This should be done in a future PR though (and after all prominent PRs are merged to avoid conflicts).

Whatever style choice is fine with me, I just want the freaking tools to do what they're supposed to. I've used `clang-format` on C code for like 8 years now, and haven't had 1/100th the issues that Prettier has given me.

**Comment by Calinou** [comment]
> I've used clang-format on C code for like 8 years now, and haven't had 1/100th the issues that Prettier has given me.

This is one of the reasons I prefer using *just* ESLint instead of ESLint + Prettier. Less tools means less moving parts that can potentially break :slightly_smiling_face:

**Comment by DaelonSuzuka** [comment]
Honestly, the entire type/javascript experience has been right in line with trying to make Prettier behave. There's some stuff that I like, but *boy* am I glad I'm not a web developer...

**Comment by DaelonSuzuka** [comment]
@Calinou Are there actually still changes requested? I'm pretty sure I resolved everything that I saw.

**Comment by Nukiloco** [comment]
What is the current progress on this PR?

**Comment by DaelonSuzuka** [comment]
> What is the current progress on this PR? This seems like a great feature for Godot vscode!

@Nukiloco Sorry for the radio silence here. I think everybody's just busy, either with Godot 4 Beta activity or other life stuff. I really appreciate your interest here, and hope to get the active PRs cleaned up soon.

> Also does this PR also do the inspector, or is this just only for the scene?

Can you rephrase the question here? I'm not sure what you mean by "inspector".

**Comment by atirut-w** [comment]
> Can you rephrase the question here? I'm not sure what you mean by "inspector".

I think he meant node inspectors. You know, where you changer node properties like position, exported var, etc.?

**Review by Calinou — APPROVED** [review]
Thanks!

**Comment by AlfishSoftware** [comment]
> I very desperately wanted to put icons on the far right of each tree item to indicate it's visibility, whether it has a script attached, if it's been marked unique, etc etc.

Have you tried using emoji? If text is supported, it might work.
For example, you could use "📜 Node" instead of "S | Node".
Some suggestions:
- Script attached: 📜 = scroll
- Unique path: 🏷️ = label
- Locked: 🔒 = lock
- Invisible: 👁️ = eye (other options: 💭 = thought_balloon; 👻 = ghost)

**Comment by DaelonSuzuka** [comment]
@AlfishSoftware 

I did use emojis before in an earlier version, but I also need (some of) them to be clickable buttons. I may come back and revise this to split some tags into emojis and some into buttons.
